File: solutions/0144.Binary-Tree-Preorder-Traversal/Solution2.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
// iterative way using a stack.
let preorderTraversal = function (root) {
  let res = []
  if (root === null) return res
  let stack = []
  stack.push(root)
  let poped = null
  while (stack.length > 0) {
    poped = stack.pop()
    res.push(poped.val)
    if (poped.right != null) { stack.push(poped.right) }
    if (poped.left != null) { stack.push(poped.left) }
  }
  return res
}

File: solutions/0144.Binary-Tree-Preorder-Traversal/Solution2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# iterative way using a stack.


class Solution:
    def preorderTraversal(self, root):
        res = []
        if root is None:
            return res
        stack = []
        stack.append(root)
        while len(stack) > 0:
            poped = stack.pop()
            res.append(poped.val)
            if poped.right is not None:
                stack.append(poped.right)
            if poped.left is not None:
                stack.append(poped.left)
